---
title: "Individual write up"
author: "Siwen Wang"
format: html
editor: visual
---

# Load data

## Setup packages and data check

```{r, warning=FALSE, message=FALSE}
#if (!requireNamespace("BiocManager", quietly = TRUE))
#    install.packages("BiocManager")
#BiocManager::install("rnaseqGene", version = "3.8")
#BiocManager::install("DESeq2")
library(DESeq2) # for RNA-seq analysis
library(dplyr) # for data processing
library(magrittr)
library(apeglm) # Used by DESeq2 to fit a linear model
library(ggplot2) # for visualization
library(vsn) # for variance stabilization
library(hexbin) # for visualization
#BiocManager::install("edgeR")
library(edgeR)
library(gplots)
library(Biobase)
library(limma)
library(knitr)

#Load
filtered_eset <- readRDS("luad_gtex_eset_1207.rds")
#investigate
filtered_eset
```

```{r}
# Extract counts and sample metadata
countdata <- assayData(filtered_eset)$counts   # genes x samples
coldata   <- pData(filtered_eset)             # samples x variables

# Create a group variable: normal (GTEx) vs tumor (TCGA)
coldata$group <- ifelse(coldata$dataset == "TCGA_LUAD", "tumor", "normal")
coldata$group <- factor(coldata$group, levels = c("normal", "tumor"))
table(coldata$group)

```

## Prepare sex and age variables

```{r}
#handle covariates - sex
coldata$sex <- ifelse(
  !is.na(coldata$gtex.sex), 
  coldata$gtex.sex, 
  coldata$tcga.cgc_case_gender
)

coldata <- as.data.frame(coldata)
coldata$sex <- as.character(coldata$sex)
# Recode GTEx: 1 -> MALE, 2 -> FEMALE
is_gtex <- coldata$dataset == "GTEx_LUNG"
coldata$sex[is_gtex & coldata$sex %in% c("1", 1)] <- "MALE"
coldata$sex[is_gtex & coldata$sex %in% c("2", 2)] <- "FEMALE"

coldata$sex <- factor(coldata$sex, levels = c("FEMALE", "MALE"))
table(coldata$sex, coldata$dataset)
```

```{r}
#colnames(pData(filtered_eset)[grep("age", colnames(pData(filtered_eset)), ignore.case = TRUE)])
#table(coldata$gtex.age)
#summary(coldata$tcga.cgc_case_age_at_diagnosis)

# TCGA ages 
tcga_age <- coldata$tcga.cgc_case_age_at_diagnosis

# Create comparable age categories
tcga_age_bins <- cut(
  tcga_age,
  breaks = c(20, 30, 40, 50, 60, 70, 80, 90),
  labels = c("20-29", "30-39", "40-49", "50-59", "60-69", "70-79", "80-89"),
  right = FALSE
)

# Assign only to TCGA samples
tcga_age_bins[ coldata$dataset != "TCGA_LUAD" ] <- NA

coldata$age_bin <- ifelse(
  coldata$dataset == "GTEx_LUNG",
  coldata$gtex.age,
  as.character(tcga_age_bins)
)

coldata$age_bin <- factor(coldata$age_bin,
                          levels = c("20-29","30-39","40-49","50-59","60-69","70-79", "80-89"))

table(coldata$age_bin, coldata$dataset)
```

```{r}
#check numbers
table(coldata$dataset, is.na(coldata$sex))
table(coldata$dataset, is.na(coldata$age_bin))
```

```{r, message=FALSE}
# drop 20 samples with missing covariates
vars_for_design <- c("sex", "age_bin", "group")

keep <- complete.cases(coldata[, vars_for_design])
table(keep)  

coldata_filt   <- coldata[keep, ]
countdata_filt <- countdata[, keep]

# sanity check
stopifnot(all(colnames(countdata_filt) == rownames(coldata_filt)))

dds <- DESeqDataSetFromMatrix(
  countData = round(countdata_filt),
  colData   = coldata_filt,         
  design    = ~ sex + age_bin + group
)

dds
nrow(dds)
# saveRDS(dds, file = "data/dds.rds") # export for Emily
```

```{r}
# Generate Table 1
table(coldata_filt$dataset, coldata_filt$sex)
table(coldata_filt$dataset, coldata_filt$age_bin)

bin_medians <- c(
  "20-29" = 25,
  "30-39" = 35,
  "40-49" = 45,
  "50-59" = 55,
  "60-69" = 65,
  "70-79" = 75,
  "80-89" = 85
)

coldata$age_num <- bin_medians[ as.character(coldata$age_bin) ]
age_summary_by_group <- coldata %>%
  group_by(group) %>%
  summarise(
    n = sum(!is.na(age_num)),
    mean_age = mean(age_num, na.rm = TRUE),
    sd_age = sd(age_num, na.rm = TRUE)
  )
kable(
  age_summary_by_group,
  caption = "Mean (SD) Age by Tumor Status"
)
```

# DESeq2

## Pre-filtering datasets

```{r}
#rows with 2 or more counts across all samples
dds <- dds[ rowSums(counts(dds)) > 1, ]
nrow(dds)
```

```{r, warning=FALSE, message=FALSE}
lambda <- 10^seq(from = -1, to = 2, length = 1000)
cts <- matrix(rpois(1000*100, lambda), ncol = 100)
library("vsn") # for variance stabilization
library("hexbin") # for visualization
meanSdPlot(cts, ranks = FALSE)
log.cts.one <- log2(cts + 1)
meanSdPlot(log.cts.one, ranks = FALSE)
```

```{r, warning=FALSE, message=FALSE}
#Use the variance stabilizing transformation (VST) for count data that stabilize the variance across the mean: 
vsd <- vst(dds, blind = FALSE)
head(assay(vsd), 3)

dds <- estimateSizeFactors(dds)

df <- bind_rows(
  as_data_frame(log2(counts(dds, normalized=TRUE)[, 1:2]+1)) %>%
         mutate(transformation = "log2(x + 1)"),
  as_data_frame(assay(vsd)[, 1:2]) %>% mutate(transformation = "vst"))
  
colnames(df)[1:2] <- c("x", "y")  
ggplot(df, aes(x = x, y = y)) + geom_hex(bins = 80) +
  coord_fixed() + facet_grid( . ~ transformation) 
```

## Analysis

```{r}
dds <- DESeq(dds)
res <- results(dds)
res
```

```{r}
summary(res)
```

## Summarizing results

```{r}
#raw results
sum(res$pvalue < 0.05, na.rm=TRUE)
sum(!is.na(res$pvalue)) 
```

```{r}
#FDR correction
res.05 <- results(dds, alpha = 0.05)
table(res.05$padj < 0.05)
```

```{r}
#Raise the log2 fold change threshold to 2
resLFC2 <- results(dds, lfcThreshold=2)
table(resLFC2$padj < 0.05)
```

```{r}
sum(res$padj < 0.1, na.rm=TRUE)
resSig <- subset(res, padj < 0.1)

#strongest down-regulation
head(resSig[ order(resSig$log2FoldChange), ])
#strongest up-regulation
head(resSig[ order(-resSig$log2FoldChange), ])
```

## Plotting results

```{r}
res.noshr <- results(dds, name="group_tumor_vs_normal")
DESeq2::plotMA(res.noshr, ylim = c(-10, 10))
```

```{r}
# A strict threshold that with shrinkage
res_shrunk <- lfcShrink(dds, coef = "group_tumor_vs_normal", type = "apeglm")

res_shrunk_strict <- res_shrunk[
  !is.na(res_shrunk$padj) &
    res_shrunk$padj < 0.05 &
    abs(res_shrunk$log2FoldChange) > 2,
]

DESeq2::plotMA(res_shrunk_strict, ylim = c(-10, 10))
```

```{r}
hist(res$pvalue[res$baseMean > 1], breaks = 0:20/20,
     col = "grey50", border = "white",main='P-value histogram')
```

## Export results

```{r}
de_genes <- rownames(res)[ which(res$padj < 0.1 & !is.na(res$padj)) ]
length(de_genes)

de_eset <- filtered_eset[de_genes, ]

# Add DESeq2 statistics into featureData
de_stats <- as.data.frame(res[de_genes, ])
fData(de_eset) <- cbind(fData(de_eset), de_stats)

saveRDS(de_eset, file = "data/DE_genes_only_eset_DESeq2.rds")
```

# Limma-voom

## Analysis

```{r, warning=FALSE, message=FALSE}
countdata_filt <- countdata[, keep]
coldata_filt   <- coldata[keep, ]
group          <- coldata_filt$group

y <- DGEList(counts = countdata_filt)

# filtering using the same for DESeq2 pre-filtering
keep_genes2 <- filterByExpr(y, group = group)
y <- y[keep_genes2, , keep.lib.sizes = FALSE]
y <- calcNormFactors(y)

design_limma <- model.matrix(~ sex + age_bin + group, data = coldata_filt)
```

```{r}
v <- voom(y, design_limma, plot = TRUE)
fit_limma <- lmFit(v, design_limma)
fit_limma <- eBayes(fit_limma)

res_limma <- topTable(fit_limma, coef = "grouptumor", number = Inf)

topTable(fit_limma, coef = "grouptumor", n = 20)  
```

## Plot results

```{r}
plotSA(fit_limma, main="Final model mean-variance trend")
```

## Plotting results

```{r}
#Explore results
sum(res_limma$adj.P.Val < 0.05)
```

```{r}
# Find the number of genes differentially expressed under a defined thereshold (FDR<0.05 and absolute log fold-change greater than 2).
sum(res_limma$adj.P.Val < 0.05 & abs(res_limma$logFC) > 2)
```

```{r}
# Find the number of genes overexpressed in the cell lines
sum(res_limma$adj.P.Val < 0.05 & res_limma$logFC > 2)
```

```{r}
# Find the number of genes overexpressed in the tissue samples
sum(res_limma$adj.P.Val < 0.05 & res_limma$logFC < -2)
```

```{r}
# Volcano-plot
# Color genes differentially expressed under a defined thereshold (FDR<0.05 and absolute log fold-change greater than 2).
sig = ifelse(res_limma$adj.P.Val < 0.05 & abs(res_limma$logFC) > 2, yes="red",no="black")
plot(x=res_limma$logFC, y=-log(res_limma$adj.P.Val), xlab="log-fold-change",
     ylab="-log(FDR)",pch=21,col=sig,bty="l", main="Volcano plot", cex=0.8)
```

```{r}
# MA plot
sig = ifelse(res_limma$adj.P.Val < 0.05 & abs(res_limma$logFC) > 2, yes="red",no="black")
plot(x=res_limma$AveExpr, y=res_limma$logFC, xlab="Average log-expression", ylab="log-fold-change",
     pch=21,col=sig,bty="l", main="MA plot", cex=0.8)
```

```{r, message=FALSE, warning=FALSE}
# Heatmap with top 20 differentially expressed genes
y <- DGEList(counts = countdata_filt)
y <- calcNormFactors(y)

logcpm0 <- log2(cpm(y) + 1)   # forces min = 0

res_filt <- subset(res_limma, adj.P.Val < 0.05 & abs(logFC) > 2)
top_genes <- rownames(res_filt)[1:20]
#mat <- v$E[top_genes, ]
mat <- logcpm0[top_genes, ]

# Create color bar with group information
cols <- coldata_filt$group
heatmapColColors <- c("blue", "red")[factor(cols)]
# Create a vector for the heatmap color
heatmapCols = colorRampPalette(c("yellow", "red"))(250)
# Plot the heatmap and legend
heatmap.2(
  mat,
  trace = "none",
  ColSideColors = heatmapColColors,
  col = heatmapCols,
  labCol = FALSE,
  margins = c(1, 12),
  density.info = "none",
  key.xlab = "expression", 
  key.title = NA
)

legend(
  "topright",
  legend = levels(factor(cols)),
  fill = c("blue", "red"),
  xpd = TRUE,
  box.lwd = NA
)
```

```{r}
# Replace row names by gene symbol
feat <- fData(filtered_eset)
symbol <- feat[rownames(mat), "gene_name"]
heatmap.2(
  mat,
  trace = "none",
  ColSideColors = heatmapColColors,
  col = heatmapCols,
  labCol = FALSE,
  labRow = symbol,
  margins = c(1, 12),
  density.info = "none",
  key.xlab = "expression",
  key.title = NA
)

legend(
  "topright",
  legend = unique(cols),
  fill = unique(heatmapColColors),
  xpd = TRUE,
  box.lwd = NA
)
```

## Export results

```{r}
# FDR < 0.1, like with DESeq2
res_limma_sig <- res_limma[res_limma$adj.P.Val < 0.1, ]
nrow(res_limma_sig)  

res_limma_sig$gene_id     <- rownames(res_limma_sig)
res_limma_sig$gene_symbol <- feat[res_limma_sig$gene_id, "gene_name"]  

saveRDS(res_limma_sig, file = "data/DE_genes_limma.rds")
```

# Compare results

## Main results - padj \< 0.1

```{r}
sig_deseq <- res[de_genes, ]

common_genes <- intersect(rownames(sig_deseq), rownames(res_limma_sig))
length(common_genes)

cor(sig_deseq[common_genes, "log2FoldChange"],
    res_limma_sig[common_genes, "logFC"],
    use = "complete.obs")
```

```{r}
#Venn Diagram
deseq_genes <- rownames(sig_deseq)
limma_genes <- rownames(res_limma_sig)

only_deseq  <- setdiff(deseq_genes, limma_genes)
only_limma  <- setdiff(limma_genes, deseq_genes)
both        <- intersect(deseq_genes, limma_genes)

length(deseq_genes)   # total DESeq2 sig
length(limma_genes)   # total limma sig
length(both)          # overlap
length(only_deseq)    # DESeq2-only
length(only_limma)    # limma-only

all_genes <- union(rownames(res), rownames(res_limma))

venn_mat <- cbind(
  DESeq2 = all_genes %in% deseq_genes,
  limma  = all_genes %in% limma_genes
)
rownames(venn_mat) <- all_genes

vennDiagram(venn_mat, main = "DESeq2 vs limma-voom, padj < 0.1")
```

## More stringent results - padj \< 0.05 & log2 \> 2

```{r}
sig_deseq_strict <- subset(
  res,
  padj < 0.05 & !is.na(padj) & abs(log2FoldChange) > 2
)
nrow(sig_deseq_strict)

res_limma_strict <- subset(
  res_limma,
  adj.P.Val < 0.05 & !is.na(adj.P.Val) & abs(logFC) > 2
)
nrow(res_limma_strict)

deseq_genes_strict <- rownames(sig_deseq_strict)
limma_genes_strict <- rownames(res_limma_strict)

common_genes_strict <- intersect(deseq_genes_strict, limma_genes_strict)
length(common_genes_strict)   

cor(
  sig_deseq_strict[common_genes_strict, "log2FoldChange"],
  res_limma_strict[common_genes_strict, "logFC"],
  use = "complete.obs"
)

all_genes_strict <- union(deseq_genes_strict, limma_genes_strict)

venn_mat_strict <- cbind(
  DESeq2 = all_genes_strict %in% deseq_genes_strict,
  limma  = all_genes_strict %in% limma_genes_strict
)
rownames(venn_mat_strict) <- all_genes_strict

vennDiagram(venn_mat_strict, main = "DESeq2 vs limma-voom (FDR<0.05 & |log2FC|>2)")

```

# Rerun PCA and cluster analysis

## PCA DEA genes - padj \< 0.1

```{r}
expr_mat <- assayData(filtered_eset)$log_tpm  
expr_mat <- as.matrix(expr_mat)
deg_ids <- rownames(de_eset)   
deg_ids <- intersect(deg_ids, rownames(expr_mat))

expr_deg <- expr_mat[deg_ids, keep, drop = FALSE]

# scale by gene
expr_deg_scaled <- t(scale(t(expr_deg)))

# PCA
pca_deg <- prcomp(t(expr_deg_scaled), center = FALSE, scale. = FALSE)

df_pca <- data.frame(
  PC1 = pca_deg$x[, 1],
  PC2 = pca_deg$x[, 2],
  group = coldata_filt$group   # same samples as 'keep'
)

ggplot(df_pca, aes(PC1, PC2, color = group)) +
  geom_point(size = 3, alpha = 0.8) +
  theme_bw(base_size = 14) +
  ggtitle("PCA using DESeq2 DEGs only, padj < 0.1") +
  xlab(paste0("PC1 (", round(summary(pca_deg)$importance[2,1] * 100, 1), "%)")) +
  ylab(paste0("PC2 (", round(summary(pca_deg)$importance[2,2] * 100, 1), "%)"))

```

## Cluster analysis - padj \< 0.1

```{r}
mydata <- t(expr_deg_scaled)
wss <- (nrow(mydata)-1)*sum(apply(mydata,2,var))
for (i in 2:15) wss[i] <- sum(kmeans(mydata, 
  	centers=i)$withinss)
plot(1:15, wss, type="b", xlab="Number of Clusters",
  ylab="Within groups sum of squares")
```

```{r}
# K-Means Cluster Analysis
fit <- kmeans(mydata, 6) # 6-cluster solution
# get cluster means 
aggregate(mydata,by=list(fit$cluster),FUN=mean)
# append cluster assignment
mydata <- data.frame(mydata, fit$cluster)

cluster_df <- data.frame(
  sample  = rownames(mydata),
  cluster = factor(fit$cluster)
)

ggplot(cluster_df, aes(x = cluster)) +
  geom_bar(fill = "steelblue") +
  theme_bw() +
  labs(title = "Sample Count per K-means Cluster (k = 6)",
       x = "Cluster", y = "Number of Samples")
```

```{r}
df_pca$cluster <- factor(fit$cluster)

ggplot(df_pca, aes(PC1, PC2, color = cluster)) +
  geom_point(size=3) +
  theme_bw() +
  labs(title = "K-means Clusters Shown on PCA Space")
```

## PCA DEA genes - padj \< 0.05 & log2 \> 2

```{r}
deg_ids_strict <- rownames(sig_deseq_strict)   
deg_ids_strict <- intersect(deg_ids_strict, rownames(expr_mat))

expr_deg_strict <- expr_mat[deg_ids_strict, keep, drop = FALSE]
expr_deg_strict_scaled <- t(scale(t(expr_deg_strict)))

pca_deg_strict <- prcomp(t(expr_deg_strict_scaled), center = FALSE, scale. = FALSE)

df_pca_strict <- data.frame(
  PC1 = pca_deg_strict$x[, 1],
  PC2 = pca_deg_strict$x[, 2],
  group = coldata_filt$group
)

ggplot(df_pca_strict, aes(PC1, PC2, color = group)) +
  geom_point(size = 3, alpha = 0.8) +
  theme_bw(base_size = 14) +
  ggtitle("PCA using DESeq2 DEGs only, padj < 0.05 & log2 > 2") +
  xlab(paste0("PC1 (", round(summary(pca_deg_strict)$importance[2,1] * 100, 1), "%)")) +
  ylab(paste0("PC2 (", round(summary(pca_deg_strict)$importance[2,2] * 100, 1), "%)"))

```

## Cluster analysis - padj \< 0.05 & log2 \> 2

```{r}
mydata <- t(expr_deg_strict_scaled)
wss <- (nrow(mydata)-1)*sum(apply(mydata,2,var))
for (i in 2:15) wss[i] <- sum(kmeans(mydata, 
  	centers=i)$withinss)
plot(1:15, wss, type="b", xlab="Number of Clusters",
  ylab="Within groups sum of squares")
```

```{r}
# K-Means Cluster Analysis
fit <- kmeans(mydata, 4) # 4-cluster solution
# get cluster means 
aggregate(mydata,by=list(fit$cluster),FUN=mean)
# append cluster assignment
mydata <- data.frame(mydata, fit$cluster)

cluster_df <- data.frame(
  sample  = rownames(mydata),
  cluster = factor(fit$cluster)
)

ggplot(cluster_df, aes(x = cluster)) +
  geom_bar(fill = "steelblue") +
  theme_bw() +
  labs(title = "Sample Count per K-means Cluster (k = 4)",
       x = "Cluster", y = "Number of Samples")
```

```{r}
df_pca_strict$cluster <- factor(fit$cluster)

ggplot(df_pca_strict, aes(PC1, PC2, color = cluster)) +
  geom_point(size=3) +
  theme_bw() +
  labs(title = "K-means Clusters Shown on PCA Space, padj < 0.05 & log2 > 2")
```

# Summary of results

We performed differential expression analysis (DEA) using both DESeq2 and limma-voom, adjusting for age and sex as covariates. Smoking status was excluded because it was available only for tumor samples and not for controls. Both pipelines produced broadly comparable results: a large number of genes were identified as differentially expressed after multiple-testing correction. As expected, applying a more stringent log2​ fold-change threshold resulted in a smaller, more conservative set of significant genes.

Both methods revealed a system-wide shift in gene expression between tumor (TCGA-LUAD) and normal (GTEx lung) samples. However, this pattern likely reflects systematic differences between data sources—such as technical, procedural, or cohort-level variation—rather than purely biological differences. Because the dataset indicator (TCGA vs. GTEx) is perfectly colinear with case status (tumor vs. normal), batch effect and phenotype cannot be disentangled. For this reason, both DESeq2 and limma-voom were run using raw counts rather than batch-corrected expression values. Correcting for dataset batch would have removed the tumor–normal contrast entirely.

PCA restricted to DEGs also showed separation similar to the global PCA, suggesting that most differential expression signal reflects dataset-specific variation rather than tumor-specific transcriptional changes. The clustering analyses also failed to detect meaningful clusters due to this data limitation.

Interpretation of the results should therefore be made with caution, as significant findings may partially reflect inter-dataset technical differences rather than true tumor-specific transcriptional changes.
