---
title: "test code"
format: html
editor: visual
---

## 1. For TCGA files

All code was from the 02_expressionSet_example.Rmd file in the class introduction R canvas folder in the posit cloud. The workflow was just adapted for TCGA files instead of the txt files that they used to demo.

Loading TCGA file

```{r}
# run the load data rmd they provided that i adapted for TCGA
library(knitr)

# Extract R code from the Rmd to a .R file
purl("load_TCGA.Rmd", output = "load_TCGA.R")

# Source the extracted R script
source("load_TCGA.R")
```

Constructing an expressionset from TCGA files

```{r}
library(Biobase)
library(limma)
library(biomaRt)
library(ggfortify)

# Convert matrix to data.frame first
pheno <- clinical[match(colnames(normalized_exp),rownames(clinical)),]
all(colnames(normalized_exp) == rownames(pheno))

pheno<-as.data.frame(pheno)
# Make an ExpressionSet
eset_tcga <- ExpressionSet(assayData= normalized_exp, phenoData=AnnotatedDataFrame(pheno))
eset_tcga
```

Adding more information to the eset: a normalized expression matrix, and the chromosome location for each gene.

```{r}
# CPM (counts per million) normalization - counts scaled by the total number of reads
cpm <- apply(exprs(eset_tcga),2, function(x) x/sum(x)*10^6)
log_cpm <- log2(cpm+1)

storageMode(eset_tcga) <- "environment"
assayData(eset_tcga)[["log_cpm"]] <- log_cpm
```

Extracting chromosome location using bioMart

```{r}
genes <- featureNames(eset_tcga)
genes_clean <- sub("\\..*", "", genes)

# 2. Connect to Ensembl (use appropriate version if needed)
ensembl <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")

# 3. Get gene mappings
gene_mapping <- getBM(
  attributes = c("ensembl_gene_id", "hgnc_symbol"),
  filters = "ensembl_gene_id",
  values = genes_clean,
  mart = ensembl
)

annot <- read.delim("class_IntroductionR_canvas/biomart_annotation_expressionSetTutorial.txt",stringsAsFactors = F)

head(annot)
```

```{r}
chr <- annot[match(gene_mapping$hgnc_symbol, annot$hgnc_symbol), "chromosome_name"]
head(chr)
```

Add features with chromosome names to eset

```{r}
featureData(eset_tcga) <- AnnotatedDataFrame(as.data.frame(chr))
eset_tcga
```

## 2. For GTEx files

```{r}
# run the load data rmd they provided that i adapted for TCGA
library(knitr)

# Extract R code from the Rmd to a .R file
purl("load_gtex.Rmd", output = "load_gtex.R")

# Source the extracted R script
source("load_gtex.R")
```

```{r}
# Convert matrix to data.frame first
pheno_gtex <- as.data.frame(clinical[match(colnames(normalized_exp_gtex),rownames(clinical_gtex)),])
all(colnames(normalized_exp_gtex) == rownames(pheno_gtex))

pheno<-as.data.frame(pheno_gtex)
# Make an ExpressionSet
eset_gtex <- ExpressionSet(assayData= normalized_exp_gtex, phenoData=AnnotatedDataFrame(pheno_gtex))
eset_gtex
```

## Sample PCA for TCGA before batch correction

```{r}
library(ggfortify)

filtered_eset<-readRDS("filtered_eset.rds")

myPCA <- prcomp(t(assayData(filtered_eset)[["log_tpm"]]))

autoplot(PCA_filter_prebatch, data=pData(filtered_eset), color="dataset", main="Before batch correction")
```

## PCA after batch correction

```{r}
Platform <- pData(eset)$tcga.cgc_case_tumor_status

valid_samples <- !is.na(Platform)

exp_clean <- assayData(eset)[["log_cpm"]][, valid_samples]
data_clean <- Platform[valid_samples]
pheno_clean <- pData(eset)[valid_samples, ]

exp_batchRem <- removeBatchEffect(exp_clean, batch = data_clean)

# exp_batchRem <- removeBatchEffect(assayData(eset)[["log_cpm"]], batch=Platform)
myPCA2 <- prcomp(t(exp_batchRem))
autoplot(myPCA2, data=pheno_clean, colour="tcga.cgc_case_tumor_status", main="After batch correction")
```

##DE with deseq2

```{r}
library(DESeq2)
# Raw counts matrix
counts_mat <- assayData(filtered_eset)[["counts"]]

# Make sure counts are integers (DESeq2 requires integers)
counts_mat <- round(counts_mat)

# Sample information
colData <- pData(filtered_eset)

# Optional: subset or rename columns to use in the design
# For example, comparing TCGA vs GTEx
colData$group <- factor(colData$dataset)
```

create DDS dataset

```{r}
dds <- DESeqDataSetFromMatrix(
  countData = counts_mat,
  colData   = colData,
  design    = ~ group   # replace 'group' with your variable of interest
)

#filter low count genes
dds <- dds[rowSums(counts(dds)) > 1, ]

# run deseq
dds <- DESeq(dds)
```

```{r}
# Default contrast: second level vs first level
res <- results(dds)

# If you want a specific contrast (e.g., GTEx vs TCGA)
res <- results(dds, contrast = c("group", "GTEx_LUNG", "TCGA_LUAD"))

# Check the variables in the model
resultsNames(dds)
```

```{r}
# Remove NA padj
res_filt <- res[!is.na(res$padj),]

# Genes upregulated in GTEx
res_up <- res_filt[res_filt$padj < 0.01 & res_filt$log2FoldChange > 1, ]
# Genes downregulated in GTEx (or up in TCGA)
res_down <- res_filt[res_filt$padj < 0.01 & res_filt$log2FoldChange < -1, ]
```

top 200 genes

```{r}
# Top 200 upregulated
top_up <- res_up[order(res_up$log2FoldChange, decreasing = TRUE), ]
top_up <- head(top_up, 200)
write.table(rownames(top_up), "top_up_genes.txt", sep="\t", row.names=FALSE, quote=FALSE)

# Top 200 downregulated
top_down <- res_down[order(res_down$log2FoldChange), ]
top_down <- head(top_down, 200)
write.table(rownames(top_down), "top_down_genes.txt", sep="\t", row.names=FALSE, quote=FALSE)

saveRDS(top_down, "top_down.rds")
saveRDS(top_up, "top_up.rds")

```
